crates/astria-sequencer/src/ibc/recover_client.rs

```diff
@@ -0,0 +1,124 @@
+            "only the sudo address can recover a client",
+        );
+
+        let timestamp = state
+            .get_block_timestamp()
+            .await
+            .wrap_err("failed to get timestamp")?;
+        let subject_client_status = state
+            .get_client_status(&self.subject_client_id, timestamp)
+            .await;
+
+        ensure!(
+            subject_client_status != ClientStatus::Active,
+            "cannot recover an active client",
+        );
+
+        let substitute_client_status = state
+            .get_client_status(&self.substitute_client_id, timestamp)
+            .await;
+
+        ensure!(
+            substitute_client_status == ClientStatus::Active,
+            "substitute client must be active: status is {}",
+            substitute_client_status,
+        );
+
+        let mut subject_client_state = state
+            .get_client_state(&self.subject_client_id)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("subject client state not found")?;
+        let substitute_client_state = state
+            .get_client_state(&self.substitute_client_id)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("substitute client state not found")?;
+
+        ensure!(
+            subject_client_state.latest_height() < substitute_client_state.latest_height(),
+            "substitute client must have a higher height than that of subject client; subject \
+             client height: {}, substitute client height: {}",
+            subject_client_state.latest_height(),
+            substitute_client_state.latest_height(),
+        );
+
+        let height = ibc_types::core::client::Height {
+            revision_height: state
+                .get_block_height()
+                .await
+                .wrap_err("failed to get block height")?,
+            revision_number: state
+                .get_revision_number()
+                .await
+                .wrap_err("failed to get revision number")?,
+        };
+        let substitute_consensus_state = state
+            .prev_verified_consensus_state(&self.substitute_client_id, &height)
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err("failed to get substitute client consensus state")?
+            .ok_or_eyre("substitute client consensus state not found")?;
+        state
+            .put_verified_consensus_state::<crate::ibc::host_interface::AstriaHost>(
+                substitute_client_state.latest_height(),
+                self.subject_client_id.clone(),
+                substitute_consensus_state,
+            )
+            .await
+            .map_err(anyhow_to_eyre)
+            .wrap_err(
+                "failed to put
+        verified consensus state",
+            )?;
+
+        subject_client_state.latest_height = substitute_client_state.latest_height;
+        subject_client_state.chain_id = substitute_client_state.chain_id;
```

Comment:
Should we actually have an ensure block on this? To make it so that recovery can never change the chain being acted on?

I suppose there may be a usecase here were there was a hardfork in chain and client expired but assets are considered the same? 

Chain id changing hardforks are rare, nearly all the time we probably want to replace with a client of the same chain id and not doing so would be a safety failure? I'd be ok just not supporting that usecase at this time, and having to evaluate what to do if it does come up?

Reply:
the spec allows the chain ID to be changed - https://ibc.cosmos.network/architecture/adr-026-ibc-client-recovery-mechanisms/ although i'm not sure exactly why. this might be for the case you said where the counterparty forks and changes their chain ID, but we didn't do a client upgrade in time, which would invalidate the client.